dan_layer/consensus/src/hotstuff/on_receive_vote.rs

```diff
@@ -1,109 +1,142 @@
-        //         let leaf_hashes = votes.iter().map(|v| v.node_hash()).collect::<Vec<_>>();
-        //
-        //         // TODO: Check all votes
-        //         let main_vote = votes.get(0).unwrap();
-        //
-        //         let qc = QuorumCertificate::new(
-        //             node.payload_id(),
-        //             node.payload_height(),
-        //             *node.hash(),
-        //             node.height(),
-        //             node.shard(),
-        //             node.epoch(),
-        //             self.public_key.clone(),
-        //             main_vote.decision(),
-        //             main_vote.all_shard_pledges().clone(),
-        //             validator_metadata,
-        //             Some(merged_proof),
-        //             leaf_hashes,
-        //         );
-        //         self.update_high_qc(&mut tx, node.proposed_by().clone(), qc)?;
-        //
-        //         on_propose = Some((node.shard(), node.payload_id()));
-        //     }
-        //
-        //     // commit the transaction
-        //     tx.commit()?;
-        // }
-        //
-        // // Propose the next node
-        // if let Some((shard_id, payload_id)) = on_propose {
-        //     // TODO: This should go in a some component that controls message flows and events
-        //     let epoch = self.epoch_manager.current_epoch().await?;
-        //     let committee = self.epoch_manager.get_committee(epoch, shard_id).await?;
-        //     if committee.is_empty() {
-        //         return Err(HotStuffError::NoCommitteeForShard { shard: shard_id, epoch });
-        //     }
-        //     if self.is_leader(payload_id, shard_id, &committee)? {
-        //         self.leader_on_propose(shard_id, payload_id).await?;
-        //     }
-        // }
-        // Ok(())
+
+        let addr = self.epoch_manager.get_our_validator_addr(message.epoch).await?;
+        let committee = self.epoch_manager.get_local_committee(message.epoch).await?;
+        if !committee.contains(&addr) {
+            return Err(HotStuffError::ReceivedMessageFromNonCommitteeMember {
+                epoch: message.epoch,
+                sender: from.to_string(),
+                context: "OnVoteReceived".to_string(),
+            });
+        }
+
+        if !self.leader_strategy.is_leader(&addr, &committee, &message.block_id, 0) {
+            return Err(HotStuffError::NotTheLeader {
+                details: format!("Not this leader for block {}, vote sent by {}", message.block_id, addr),
+            });
+        }
+
+        let local_commmittee_shard = self.epoch_manager.get_local_committee_shard(message.epoch).await?;
+
+        // Get the sender shard, and check that they are in the local committee
+        let sender_shard = self
+            .epoch_manager
+            .get_validator_shard(message.epoch, from.clone())
+            .await?;
+        if !local_commmittee_shard.includes_shard(&sender_shard) {
```

Comment:
wouldn't the local_committee_shard only be 1 shard?

Reply:
A `CommitteeShard` represents the shard space for a committee. So includes_shard will return true for any ShardId that falls within a start and end range (represented by the bucket and num_committees)

```
is_in_committee(s: ShardId) -> bool {
  committee_shard_size = u256::MAX / num_committees
  start = bucket * committee_shard_size
  end = (bucket + 1) * committee_shard_size

  return s >=start && s < end
}
```

Reply:
alright, we should standardize on terminology but doesn't need to be corrected in this pr